### PR TITLE
[LoopUnroll] Remove redundant debug instructions after blocks have been merged

### LINKED ISSUE #89073: [BasicBlockUtils]:  remove redundant llvm.dbgs intrunctions to fix compile time error.

### PR DESCRIPTION
**Pull request by coderchenlin**
Remove redundant debug instructions after blocks have been merged into the predecessor, It can reduce some compile time in some cases.

This change only fixes the situation of loop unrolling, and other situations are not considered. "RemoveRedundantDbgInstrs" seems to be very time-consuming. Thus, we just add here after the "Dest" has been merged into the "Fold",  this may be a more targeted solution!!!

fixes: https://github.com/llvm/llvm-project/issues/89073

### CODE REVIEW

**Comment by coderchenlin** [comment]
@slackito @aeubanks , Hi, guys, I'm sorry to bother you. I just give a more targeted solution for fixing compile-time. and I also test my case and your provided case, the compile-time is Ok.  Could you help me to review this change, and give some advices?



[Uploading slowdown.ll.txt…]()
[before.time-trace.json](https://github.com/llvm/llvm-project/files/15234414/before.time-trace.json)
[after.time-trace.json](https://github.com/llvm/llvm-project/files/15234415/after.time-trace.json)

**Comment by coderchenlin** [comment]
@fhahn, Apologies for your time,  Could you help me to review this change again?  As @slackito found,  the original solution will introduce quadratic behavior in SimplifyCFG and make compilation time increase. Thus, I change the location, and put  the changes in loop-unrolling, and it won't have any effects in SimplifyCFG.

File: llvm/test/Transforms/LoopUnroll/unroll-remove-redundant-dbg.ll [hunk 1]
```diff
@@ -0,0 +1,45 @@
+; RUN: opt < %s -S -passes=loop-unroll | FileCheck %s
```
Comment (aeubanks) [inline llvm/test/Transforms/LoopUnroll/unroll-remove-redundant-dbg.ll:1]: use llvm/utils/update_test_checks.py

**Review by aeubanks — COMMENTED** [review]
this seems like it would still run into a similar issue where we may repeatedly call `RemoveRedundantDbgInstrs` on some set of growing dbg instructions, if loop-unroll manages to create a bunch of mergeable blocks. can we ensure that we only call it on an instruction at most once by calling `RemoveRedundantDbgInstrs` not every time we merge a block, but on a set of merged blocks?

e.g. if we merge A->B->C, then D->E->F, we should call `RemoveRedundantDbgInstrs` on C once we see we can't merge C->D, and then we call `RemoveRedundantDbgInstrs` on F when we're done (and we saw that F did indeed have a block merged into it)

**Comment by coderchenlin** [comment]
> we should call `RemoveRedundantDbgInstrs` on C once we see we can't merge C->D, and then we call `RemoveRedundantDbgInstrs` on F when we're done (and we saw that F did indeed have a block merged into it)

Thanks for your advice, @aeubanks , I put the process in `simplifyLoopAfterUnroll`, It seems this position is more suitable after loop-unroll is finished.  The result on the case `A->B->C`, A firstly merge B, and then, merge C.

Before:
 after A merge B, then call `RemoveredundantDbginstrs`, after A merge C, continue call  `RemoveredundantDbginstrs`.

Now：
  after A merge B, and A merge C, finally, call  `RemoveredundantDbginstrs`. Only call once,

File: llvm/test/Transforms/LoopUnroll/unroll-remove-redundant-dbg.ll (hunk 1 above)
Comment (aeubanks) [reply llvm/test/Transforms/LoopUnroll/unroll-remove-redundant-dbg.ll:2]: this isn't properly using update_test_checks.py, see the comments in that file. I believe the CHECK lines below need to be removed first. as the comments in update_test_checks.py say, please verify that there is a difference in IR with and without the loop unroll cleanup change

**Review by aeubanks — COMMENTED** [review]
this makes sense, thanks! the test just needs to be fixed

File: llvm/test/Transforms/LoopUnroll/unroll-remove-redundant-dbg.ll (hunk 1 above)
Comment (coderchenlin) [reply llvm/test/Transforms/LoopUnroll/unroll-remove-redundant-dbg.ll:2]: Thanks, I use update_test_checks.py to generate CHECKs, and keep some important contents

Comment (aeubanks) [reply llvm/test/Transforms/LoopUnroll/unroll-remove-redundant-dbg.ll:2]: you shouldn't manually adjust CHECK lines generated by update_test_checks.py, just keep whatever it generates

**Review by nikic — APPROVED** [review]
LGTM

File: llvm/test/Transforms/LoopUnroll/unroll-remove-redundant-dbg.ll (hunk 1 above)
Comment (coderchenlin) [reply llvm/test/Transforms/LoopUnroll/unroll-remove-redundant-dbg.ll:2]: Thanks, keep all the  CHECK lines.

**Review by aeubanks — APPROVED** [review]
lgtm, thanks!

**Pull request merged**